ompi/mca/osc/monitoring/osc_monitoring_component.c

```diff
@@ -19,39 +19,52 @@
 #include <opal/mca/base/mca_base_component_repository.h>
 
 /***************************************/
-/* Include template generating macros */
-#include "osc_monitoring_template.h"
 
-#include <ompi/mca/osc/rdma/osc_rdma.h>
-OSC_MONITORING_MODULE_TEMPLATE_GENERATE(rdma, ompi_osc_rdma_module_t, comm)
-#undef GET_MODULE
+static opal_hash_table_t *win_to_comm_ht = NULL;
 
-#include <ompi/mca/osc/sm/osc_sm.h>
-OSC_MONITORING_MODULE_TEMPLATE_GENERATE(sm, ompi_osc_sm_module_t, comm)
-#undef GET_MODULE
+static inline struct ompi_communicator_t*
+ompi_osc_monitoring_get_comm(ompi_win_t *win)
+{
+    ompi_communicator_t *comm = NULL;
+    uint64_t key = *((uint64_t*)&win);
+    int ret = opal_hash_table_get_value_uint64(win_to_comm_ht, key,
+                                               (void *)&comm);
+    return OPAL_SUCCESS == ret ? comm : NULL;
+}
 
-#include <ompi/mca/osc/pt2pt/osc_pt2pt.h>
-OSC_MONITORING_MODULE_TEMPLATE_GENERATE(pt2pt, ompi_osc_pt2pt_module_t, comm)
-#undef GET_MODULE
+/***************************************/
+/* Include template generating macros */
+#include "osc_monitoring_template.h"
 
-#ifdef OMPI_WITH_OSC_PORTALS4
-#include <ompi/mca/osc/portals4/osc_portals4.h>
-OSC_MONITORING_MODULE_TEMPLATE_GENERATE(portals4, ompi_osc_portals4_module_t, comm)
-#undef GET_MODULE
-#endif /* OMPI_WITH_OSC_PORTALS4 */
+OSC_MONITORING_MODULE_TEMPLATE_GENERATE(rdma)
```

Comment:
This still feels pretty brittle (although it is a step up).  There probably not a generic solution, but a less-brittle version would be to have an autogenerated file from configure with these OSC_MONITORING_MODULE_TEMPLATE_GENERATE() statements, which was generated during configure time by looping over `mca_ompi_osc_m4_config_component_list` and `mca_ompi_osc_no_config_component_list`.

As an example, I notice that there's no template for the `ucx` component.

Reply:
Indeed I have not came back for a few few months, and I missed this new component. You are totally right on how careful this whole component have to be handle. I had not thought of this m4 solution, but that's a very good idea.